cmd/syncthing/gui.go

```diff
@@ -329,36 +316,70 @@ func (s *apiService) Serve() {
 	}
 
 	// Redirect to HTTPS if we are supposed to
-	if guiCfg.UseTLS() {
+	if guiCfg.UseTLS {
 		handler = redirectToHTTPSMiddleware(handler)
 	}
 
 	// Add the CORS handling
 	handler = corsMiddleware(handler, guiCfg.InsecureAllowFrameLoading)
 
-	if addressIsLocalhost(guiCfg.Address()) && !guiCfg.InsecureSkipHostCheck {
+	if addressIsLocalhost(guiCfg.Address) && !guiCfg.InsecureSkipHostCheck {
 		// Verify source host
 		handler = localhostMiddleware(handler)
 	}
 
 	handler = debugMiddleware(handler)
 
-	srv := http.Server{
-		Handler: handler,
-		// ReadTimeout must be longer than SyncthingController $scope.refresh
-		// interval to avoid HTTP keepalive/GUI refresh race.
-		ReadTimeout: 15 * time.Second,
+	return handler
+}
+
+func (s *apiService) Serve() {
+	guiCfgs := s.cfg.GUIs()
+	listeners, err := s.getListeners(guiCfgs)
+	if err != nil {
+		select {
+		case <-s.startedOnce:
+			// We let this be a loud user-visible warning as it may be the only
+			// indication they get that the GUI won't be available.
+			l.Warnln("Starting API/GUI:", err)
+			return
+
+		default:
+			// This is during initialization. A failure here should be fatal
+			// as there will be no way for the user to communicate with us
+			// otherwise anyway.
+			l.Fatalln("Starting API/GUI:", err)
+		}
+	}
+
+	if len(listeners) == 0 {
+		// Not much we can do here other than exit quickly. The supervisor
+		// will log an error at some point.
```

Comment:
To be honest, the UI could be disabled.
We didn't scream loud before, and we probably should, but it needs to happen with sanity checks that we are expected to have a UI running.

Reply:
Agreed, not something to be done in this PR. I mostly just jumped at commenting when reading this, because as far as my experience goes the suture stuff is nice in that it keeps running but (in it's current form) does hide serious problems.

In short: Nothing to be solved/discussed in this PR (if at all).